src/builder/edit_interaction_response.rs

```diff
@@ -126,7 +127,15 @@ impl Builder for EditInteractionResponse {
         ctx: Self::Context<'_>,
     ) -> Result<Self::Built> {
         self.0.check_length()?;
+
         let files = std::mem::take(&mut self.0.files);
+        if !files.is_empty() {
```

Comment:
Why this check?

Reply:
If the payload includes the `attachments` field, all existing attachments but the ones specified there are removed. If we unconditionally create the vector, the `attachments` field is not skipped during serialization and added as an empty array, meaning that all existing attachments would be always removed.

If we skip the serialization if the vector is empty, users are unable to clear all attachments using `clear_existing_attachments`.

The fact that adding new files or calling `clear_existing_attachments` or `keep_existing_attachment` (both initialize the vector, and the latter extends it) removes all existing, not manually added attachments from the message is known and (I think) documented behavior. The check is needed to keep that behavior.

Reply:
I see. All of that makes perfect sense.

This situation, where existing attachments are sometimes removed, sometimes not, depending on which methods are called, is really confusing ngl. I wonder if serenity can invent a less footgunny API here. Or maybe we can get away _always_ supplying `attachments`, but pre-filling it with AttachmentId's wherever serenity has the required data, and requiring users to do it where not?

Reply:
I wrote up all the possible combinations of `attachments` and `files`, with existing and new attachments, and what will happen to the message:

```md
If we edit a message that currently has attachments [old], then the following possible request variants exist:

attachments None, files []
-> attachments unchanged

attachments None, files [new]
-> attachments replaced

attachments Some([]), files []
-> attachments removed

attachments Some([]), files [new]
-> attachments removed (new attachments NOT ADDED!)

attachments Some([old]), files []
-> attachments unchanged, existing attachments configurable

attachments Some([old]), files [new]
-> attachments unchanged, existing attachments configurable (new attachments NOT ADDED!)

attachments Some([new]), files []
-> attachments removed

attachments Some([new]), files [new]
-> attachments replaced, new attachments configurable (e.g. `description`)

attachments Some([old, new]), files []
-> attachments unchanged, existing attachments configurable

attachments Some([old, new]), files [new]
-> attachments appended, both existing and new attachments configurable
```

So these are all possible unique actions we can take with what we have:

1. attachments removed (no extra data needed)
2. attachments unchanged (no extra data needed)
3. attachments unchanged, existing attachments configurable (**old attachment IDs needed**)
4. attachments replaced (no extra data needed)
5. attachments replaced, new attachments configurable (e.g. `description`) (no extra data needed)
6. attachments appended, both existing and new attachments configurable (**old attachment IDs needed**)

Of these, 4 is redundant, because 5 is strictly more powerful and requires no extra data

So, these remaining 5 possibilities could be represented in Rust like this

```rust
pub enum EditAttachments {
    RemoveExisting, // 1
    KeepExisting, // 2
    ConfigureExisting(Vec<ExistingAttachment>), // 3
    RemoveExistingAndAddNew(Vec<NewAttachment>), // 5
    KeepOrConfigureExistingAndAddNew(Vec<ExistingAttachment>, Vec<NewAttachment>), // 6
}
```

Technically, RemoveExisting is redundant in favor of ConfigureExisting(vec![]), RemoveExistingAndAddNew(...) is redundant in favor of KeepOrConfigureExistingAndAddNew(vec![], ...), and ConfigureExisting(...) is redundant in favor of KeepOrConfigureExistingAndAddNew(..., vec![]). So this enum can be consolidated into:

```rust
enum EditAttachments {
    KeepExisting,
    Configure {
        // Set to vec![] to remove all, set to vec![<all AttachmentIds from the msg>] to keep all
        existing: Vec<ExistingAttachment>,
        // Set to vec![] if you just want to edit existing
        new: Vec<NewAttachment>,
    }
}
```

We could go one step ahead, and say that builder structs should always store EditAttachments in an Option, and None will correspond to KeepExisting. Then the code simplifies to:

```rust
pub struct EditAttachments {
    /// Set to vec![] to remove all, set to vec![<all AttachmentIds from the msg>] to keep all
    pub existing: Vec<ExistingAttachment>,
    /// Set to vec![] if you just want to edit existing
    pub new: Vec<NewAttachment>,
}
```

Now, `NewAttachment` could also start containing the `CreateAttachment` including the file contents within itself. This would simplify the API quite a lot I think, and remove footguns

Reply:
> We could go one step ahead, and say that builder structs should always store EditAttachments in an Option, and None will correspond to KeepExisting. Then the code simplifies to:
>
> ```rust 
> pub struct EditAttachments {
>     /// Set to vec![] to remove all, set to vec![<all AttachmentIds from the msg>] to keep all
>     pub existing: Vec<ExistingAttachment>,
>     /// Set to vec![] if you just want to edit existing
>     pub new: Vec<NewAttachment>,
> }
> ```

Internally, that's pretty much how data is laid out right now, but separated and as a single struct (`Option<EditAttachments>` instead of `Option<Vec<MessageAttachment>>`). Or do you want to expose this to the users as a new builder, and remove all existing methods of the message builders?

Reply:
I think that would be easier to understand, yes. Easier to understand in the code for devs, and easier to understand for users

Reply:
I'm currently working on it locally as an experiment